File: android-agent/src/main/java/io/opentelemetry/android/ScreenAttributesSpanProcessor.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android;

import static io.opentelemetry.android.config.RumConstants.SCREEN_NAME_KEY;

import io.opentelemetry.android.internal.services.visiblescreen.VisibleScreenService;
import io.opentelemetry.context.Context;
import io.opentelemetry.sdk.trace.ReadWriteSpan;
import io.opentelemetry.sdk.trace.ReadableSpan;
import io.opentelemetry.sdk.trace.SpanProcessor;

public final class ScreenAttributesSpanProcessor implements SpanProcessor {

    private final VisibleScreenService visibleScreenService;

    public ScreenAttributesSpanProcessor(VisibleScreenService visibleScreenService) {
        this.visibleScreenService = visibleScreenService;
    }

    @Override
    public void onStart(Context parentContext, ReadWriteSpan span) {
        String currentScreen = visibleScreenService.getCurrentlyVisibleScreen();
        span.setAttribute(SCREEN_NAME_KEY, currentScreen);
    }

    @Override
    public boolean isStartRequired() {
        return true;
    }

    @Override
    public void onEnd(ReadableSpan span) {
        // nop
    }

    @Override
    public boolean isEndRequired() {
        return false;
    }
}

File: core/src/main/java/io/opentelemetry/android/ScreenAttributesSpanProcessor.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android

import io.opentelemetry.android.common.internal.SemconvCompat.Companion.map
import io.opentelemetry.android.internal.services.visiblescreen.VisibleScreenTracker
import io.opentelemetry.api.common.AttributeKey.stringKey
import io.opentelemetry.context.Context
import io.opentelemetry.kotlin.semconv.AppAttributes.APP_SCREEN_NAME
import io.opentelemetry.kotlin.semconv.IncubatingApi
import io.opentelemetry.sdk.trace.ReadWriteSpan
import io.opentelemetry.sdk.trace.ReadableSpan
import io.opentelemetry.sdk.trace.SpanProcessor

/**
 * This class appends the screen name to all spans.
 */
internal class ScreenAttributesSpanProcessor(
    private val visibleScreenTracker: VisibleScreenTracker,
) : SpanProcessor {
    @OptIn(IncubatingApi::class)
    override fun onStart(
        parentContext: Context,
        span: ReadWriteSpan,
    ) {
        val currentScreen = visibleScreenTracker.currentlyVisibleScreen
        span.setAttribute(stringKey(map(APP_SCREEN_NAME)), currentScreen)
    }

    override fun isStartRequired(): Boolean = true

    override fun onEnd(span: ReadableSpan) {
        // nop
    }

    override fun isEndRequired(): Boolean = false
}
